Native implementation of aggregate functions for FLOAT

maciejgrzybek opened the pull request:
It is to avoid confusion when coercions to DOUBLE are used and e.g. sum(float) = double.
E.g. when single value is given to sum(float) it would return double value which would not equal input when coercions are used.
Small commits added for review convenience, thus commit-by-commit approach is preferred.

Internal TD review: https://github.com/Teradata/presto/pull/242

CC: @cberner

dain commented:
I haven't followed this discussion, so I didn't really understand the pull request message.  Are you saying that after this PR is merged `sum(float)` will return return `float`, or are you saying that will return `double`?

maciejgrzybek commented:
@dain: you're right. It's not clear without bigger picture.
Let me try to clarify.
Currently, assuming we're based on coercions (FLOAT->DOUBLE), when native aggregate function does not exist, corresponding double-aggregate will be used.
So without this patch `sum(float)` invocations will effectively be equal to `sum((double) float)`, returning `double`.
This may confuse user because `sum(col)` != `v1 + v2 + ... + vn` for n = number-of-rows in table when `col` is of type FLOAT.
Simplest case: for one `value` in table `sum(col)` != `value` because `sum(col)` is of type `double` while `value` is FLOAT.

What's more, FLOAT operators (e.g. `add`) exist, so the output of chain of summations `((v1 + v2) + v3) + ...` returns FLOAT while non-native `sum` (invoked through FLOAT->DOUBLE coercion) returns DOUBLE.

Though, it's also not true that after applying this patch `sum(col)` = `((v1 + v2) + v3) + ...` because I decided to use DOUBLE accumulator internally to minimize error but I return proper type (I truncate DOUBLE to FLOAT before returning).
If we prefer consistency of `sum(col)` = `((v1 + v2) + v3) + ...` I can change accumulator to be of `FLOAT` type.

To wrap up: with this patch aggregate functions have return type consistent with input types.

dain commented:
Thanks!  That makes sense to me.

BTW, in most cases `sum(col) = ((v1 + v2) + v3) + ...` won't be true. The results of calculations over approximate numbers depends on the order of the operations, and aggregations do not guarantee ordering.  Of course this really doesn't matter in the case of the super low precision in floats.

cberner commented on presto-main/src/main/java/com/facebook/presto/operator/aggregation/ApproximateFloatPercentileAggregations.java:
```diff
@@ -0,0 +1,88 @@
+@AggregationFunction("approx_percentile")
+public class ApproximateFloatPercentileAggregations
+{
+    public static final InternalAggregationFunction FLOAT_APPROXIMATE_PERCENTILE_AGGREGATION =
```
These statics should be moved to the test code, like in https://github.com/prestodb/presto/pull/5597

cberner commented on presto-main/src/main/java/com/facebook/presto/operator/aggregation/ApproximateFloatPercentileAggregations.java:
```diff
@@ -0,0 +1,88 @@
+    @InputFunction
+    public static void input(DigestAndPercentileState state, @SqlType(StandardTypes.FLOAT) long value, @SqlType(StandardTypes.DOUBLE) double percentile)
+    {
+        ApproximateLongPercentileAggregations.input(state, value, percentile);
```
It looks like this relies on our internal representation of `FLOAT` sorting in the same order as `long`. Does IEEE 754 work that way?

maciejgrzybek replied:
You're right. I forgot about sign (first bit) and multiple representation of NaN (we preserve raw bits representation) which should be collapsed to one NaN in this case.
Nice catch!
Will rewrite it.

cberner commented on presto-main/src/main/java/com/facebook/presto/operator/aggregation/ApproximateFloatPercentileArrayAggregations.java:
```diff
@@ -0,0 +1,85 @@
+import static com.facebook.presto.testing.AggregationTestUtils.generateInternalAggregationFunction;
+
+@AggregationFunction("approx_percentile")
+public class ApproximateFloatPercentileArrayAggregations
```
same comments as above

cberner commented on presto-main/src/main/java/com/facebook/presto/operator/aggregation/FloatRegressionAggregation.java:
```diff
@@ -0,0 +1,69 @@
+    public static void regrIntercept(RegressionState state, BlockBuilder out)
+    {
+        double result = getRegressionIntercept(state);
+        if (Double.isFinite(result)) {
```
same comment

maciejgrzybek replied:
ditto

cberner commented on presto-main/src/main/java/com/facebook/presto/operator/aggregation/FloatCorrelationAggregation.java:
```diff
@@ -0,0 +1,55 @@
+    public static void corr(CorrelationState state, BlockBuilder out)
+    {
+        double result = getCorrelation(state);
+        if (Double.isFinite(result)) {
```
I think you meant to test `Float.isFinite`

maciejgrzybek replied:
No, https://github.com/prestodb/presto/pull/5607#discussion_r70129174

cberner commented:
Except for the sorting issue in approx_percentile and some other minor comments this looks good

cberner commented:
I think you missed tests for some of the new aggregations like float correllation

maciejgrzybek commented on presto-main/src/main/java/com/facebook/presto/operator/aggregation/FloatRegressionAggregation.java:
```diff
@@ -0,0 +1,69 @@
+    public static void regrSlope(RegressionState state, BlockBuilder out)
+    {
+        double result = getRegressionSlope(state);
+        if (Double.isFinite(result)) {
```
No, `getRegressionSlope` returns `double`.

cberner replied:
Yes, but then you cast the value to a float which could change a finite value into infinity. It seems like you should cast first and then check whether it's finite

maciejgrzybek commented:
@cberner I skipped tests wherever I reused entirely a double implementation.
I could have added tests to check whether the return type is `float` though.
I will do this.

maciejgrzybek commented:
@cberner Addressed comments. Added missing tests, rewritten existing once to match #5597.
I also found out that we've some potential inconsistency in existing implementation of `LongDoubleCoverterUtil`. More details in 24b78ab.

Now it's back to you! :)

cberner commented:
Looks good, but could you rebase? There are a bunch of merge conflicts

maciejgrzybek commented:
@cberner Here you got it! :)
There were conflicts because of https://github.com/prestodb/presto/pull/5597 got merged.

cberner commented:
Merged. Thanks!

mbasmanova commented:
@maciejgrzybek Mack, we are trying to understand why avg(real) returns real and not double value. You have some explanation in the PR description, but I don't quite get it. Any chance you could explain a bit more? 

```
It is to avoid confusion when coercions to DOUBLE are used and e.g. sum(float) = double.
E.g. when single value is given to sum(float) it would return double value which would not equal input when coercions are used.
```

CC: @highker @rschlussel @kaikalur @tdcmeehan

mbasmanova commented:
@kaikalur Sreeni explained this to me and I'd like to share this explanation for visibility. 

It makes sense for sum and avg aggregate functions to match their corresponding scalar operators, e.g. the following should be true:

```
sum(v) = v1 + v2 + ... + vN
avg(v) = (v1 + v2 + ... + vN) / N
```
, where `v<i>` is the i-th row of the "v" column and N is total number of rows. 

Since, real + real gives real, sum(real) should return real as well. Similarly, (real + real) / integer gives real, hence, avg(real) should return real as well.
